packages/okta-adapter/src/change_validators/group_rule_administrator.ts

```diff
@@ -13,61 +13,59 @@
 * limitations under the License.
 */
-import { ChangeValidator, getChangeData, isInstanceChange, isAdditionChange, isInstanceElement, InstanceElement, ChangeError, ReferenceExpression } from '@salto-io/adapter-api'
+import { ChangeValidator, getChangeData, isInstanceChange, isAdditionChange, InstanceElement } from '@salto-io/adapter-api'
 import { resolvePath, references as referenceUtils } from '@salto-io/adapter-utils'
 import _ from 'lodash'
+import { values as lowerDashValues } from '@salto-io/lowerdash'
+import { logger } from '@salto-io/logging'
 import { GROUP_RULE_TYPE_NAME } from '../constants'
 
+const { isDefined } = lowerDashValues
 const GROUP_ID_PATH = ['actions', 'assignUserToGroups', 'groupIds']
 const { isArrayOfRefExprToInstances } = referenceUtils
+const log = logger(module)
 
-const groupRefToGroupWithRoles = async (references: ReferenceExpression[]): Promise<InstanceElement[]> => (
-  await Promise.all(references.map(async reference => reference.getResolvedValue())))
-  .filter(isInstanceElement)
-  .filter(target => {
-    const roles = target.value?.roles
-    if (roles !== undefined && !(_.isEmpty(roles))) {
-      return true
+const getGroupsWithRoleByRuleId = async (groupRuleInstance: InstanceElement[]):
+Promise<Record<string, InstanceElement[]>> => {
+  const groupsWithRoleByRuleId = (await Promise.all(groupRuleInstance.map(async instance => {
+    const targetGroupsPath = instance.elemID.createNestedID(...GROUP_ID_PATH)
+    const TagetGroupReferences = resolvePath(instance, targetGroupsPath)
+    if (!isArrayOfRefExprToInstances(TagetGroupReferences)) {
+      log.debug('Could not find group references in %s', instance.elemID.getFullName())
+      return undefined
     }
-    return false
-  })
-
-const groupRulesToGroupsRecord = async (instances: InstanceElement[]): Promise<Record<string, InstanceElement[]>> => {
-  const record: Record<string, InstanceElement[]> = {}
-  await Promise.all(instances.map(async instance => {
-    const elemID = instance.elemID.createNestedID(...GROUP_ID_PATH)
-    const references = resolvePath(instance, elemID)
-    if (!isArrayOfRefExprToInstances(references)) {
-      return
-    }
-    const targets = await groupRefToGroupWithRoles(references)
-    if (targets !== undefined && !(_.isEmpty(targets))) {
-      record[instance.elemID.getFullName()] = targets
+    const targetGroupWithRoles = TagetGroupReferences.map(groupReference => groupReference.value).filter(
+      targetGroupInstance => !(_.isEmpty(targetGroupInstance.value?.roles))
+    )
+    if (!(_.isEmpty(targetGroupWithRoles))) {
+      return [instance.elemID.getFullName(), targetGroupWithRoles]
     }
-  }))
-  return record
+    return undefined
+  }))).filter(isDefined)
+  return Object.fromEntries(groupsWithRoleByRuleId)
 }
 
 /**
- * Verifies that Group target has no administrators.
+ * Verifies that the target groups for a GroupRule has no administrator roles.
```

Comment:
I think it's 'have' because we're talking a about group**s**
```suggestion
 * Verifies that the target groups for a GroupRule have no administrator roles.
```